snap/component.go

```diff
@@ -126,14 +129,88 @@ func (c *componentPlaceInfo) MountDescription() string {
 	return fmt.Sprintf("Mount unit for %s, revision %s", c.ContainerName(), c.compRevision)
 }
 
-// ReadComponentInfoFromContainer reads ComponentInfo from a snap component container.
-func ReadComponentInfoFromContainer(compf Container) (*ComponentInfo, error) {
+// ReadComponentInfoFromContainer reads ComponentInfo from a snap component
+// container. If snapInfo is not nil, it is used to complete the ComponentInfo
+// information about the component's implicit and explicit hooks, and their
+// associated plugs.
+func ReadComponentInfoFromContainer(compf Container, snapInfo *Info) (*ComponentInfo, error) {
 	yamlData, err := compf.ReadFile("meta/component.yaml")
 	if err != nil {
 		return nil, err
 	}
 
-	return InfoFromComponentYaml(yamlData)
+	componentInfo, err := InfoFromComponentYaml(yamlData)
+	if err != nil {
+		return nil, err
+	}
+
+	// if snapInfo is nil, then we can't complete the component info with
+	// implicit and explicit hooks, so we return the component info as is.
+	//
+	// we could technically create the hooks, but would be unable to bind plugs
+	// to them, so it is probably best to just leave them out.
+	if snapInfo == nil {
+		return componentInfo, nil
+	}
+
+	componentName := componentInfo.Component.ComponentName
+
+	component, ok := snapInfo.Components[componentName]
+	if !ok {
+		return nil, fmt.Errorf("internal error: %q is not a component for snap %q", componentName, snapInfo.RealName)
```

Comment:
is not clear to me that this is generally an internal error, I would leave that prefix out. Also this check doesn't seem tested